src/fia/fia/fia_state.rs

```diff
@@ -0,0 +1,131 @@
+    Path: HasDerivationPath,
+    Product: HasHDPublicKey,
+{
+    fn invalid_if_skipped(&self, factor_sources: &[FactorSource]) -> Vec<ID> {
+        assert!(self.supports_skipping_of_factor_sources);
+        todo!()
+    }
+
+    fn request_for(&self, factor_sources: &[FactorSource]) -> Self::DriverRequest {
+        assert_eq!(
+            factor_sources
+                .iter()
+                .map(|f| f.kind)
+                .collect::<HashSet<FactorSourceKind>>()
+                .len(),
+            1
+        );
+        let supports_skipping = self.supports_skipping_of_factor_sources;
+        let inputs: HashMap<FactorSourceID, HashMap<ID, Vec<Path>>> = HashMap::new();
+        if supports_skipping {
+            Self::DriverRequest::new_skippable(self.invalid_if_skipped(factor_sources), inputs)
+        } else {
+            Self::DriverRequest::new_unskippable(inputs)
+        }
+    }
+
+    fn parallel_request_for(
+        &self,
+        factor_sources_of_kind: &FactorSourcesOfKind,
+    ) -> Self::DriverRequest {
+        assert!(factor_sources_of_kind.kind.supports_parallelism());
+        self.request_for(&factor_sources_of_kind.factor_sources)
+    }
+
+    fn serial_request_for(&self, factor_source: &FactorSource) -> Self::DriverRequest {
+        assert!(!factor_source.kind.supports_parallelism());
+        self.request_for(&[factor_source.clone()])
+    }
+}
+
+impl<ID, Path, Product> FiaState<ID, Path, Product>
+where
+    ID: Hash,
+    Path: HasDerivationPath,
+    Product: HasHDPublicKey,
+{
+    pub(super) fn new(
+        supports_skipping_of_factor_sources: bool,
+        factor_sources: Vec<FactorSource>,
+    ) -> Self {
+        Self {
+            supports_skipping_of_factor_sources,
+            phantom_id: PhantomData,
+            phantom_path: PhantomData,
+            phantom_product: PhantomData,
+            factor_sources,
+        }
+    }
+
+    pub(super) type DriverRequest = BatchUseFactorSourceRequest<ID, Path>;
+    pub(super) type DriverResponse = BatchUseFactorSourceResponse<ID, Product>;
+
+    pub(super) fn next_factor_sources(&self) -> Option<FactorSourcesOfKind> {
+        todo!()
+    }
+
+    pub(super) fn handle_outcome(
+        &self,
+        response: UseFactorsAction<Self::DriverResponse>,
+    ) -> Result<()> {
+        // Mutate using interior mutability
+        todo!()
+    }
+
+    pub(super) fn is_done_early(&self) -> bool {
+        if !self.supports_skipping_of_factor_sources {
+            // Cannot skip any factor sources, so cannot be done "early".
+            return false;
+        }
+        todo!()
+    }
+
+    pub(super) fn accumulated_output(&self) -> Result<FiaOutput<ID, Product>> {
+        todo!()
+    }
+
+    pub(super) async fn reduce(
+        &self,
+        driver: &dyn UseFactorSourceDriver<ID, Path, Product>,
+        factor_sources_of_kind: &FactorSourcesOfKind,
+    ) -> Result<()> {
+        if factor_sources_of_kind.kind.supports_parallelism() {
```

Comment:
the difference is subtle! But yet very important! For FactorSourceKinds which does not support parallelism (none except DeviceFactorSource kind do! Or well pending thumbs up from @micbakos-rdx  (see comment https://github.com/Sajjon/one-does-not-simply-sign/pull/9/files#r1711372975 ) ), we MUST lazily create the requests, since if user skipped factor source `A`, that affects whether or not factor source `B` is skippable without failing any TX. Therefor we must call `handle_outcome` (contains case: `Skipped` or `Used`) on factor source `A` **before** proceeding with factor source `B` (creating the request for it).